src/vsr/grid.zig

```diff
@@ -266,45 +270,91 @@ pub fn GridType(comptime Storage: type) type {
             grid.cache.deinit(allocator);
             grid.blocks_missing.deinit(allocator);
 
-            grid.free_set_encoded.deinit(allocator);
+            grid.free_set_checkpoint.deinit(allocator);
             grid.free_set.deinit(allocator);
 
             grid.* = undefined;
         }
 
-        /// Checkpointing the grid:
-        /// 1. Writes the free set to disk.
-        /// 2. Awaits all pending repair-writes to blocks that were just freed. This guarantees that
-        ///    there are no outstanding writes to (now-)free blocks when we enter the new checkpoint.
-        /// 3. Marks blocks released during the current checkpoint as free and eligible for
-        ///    acquisition in the next checkpoint.
-        /// 4. Marks the free set's own blocks as released (but not yet free).
+        pub fn open(grid: *Grid, callback: *const fn (*Grid) void) void {
+            assert(grid.callback == .none);
+
+            grid.callback = .{ .open = callback };
+            grid.free_set_checkpoint.open(
+                grid,
+                grid.superblock.working.free_set_reference(),
+                free_set_open_callback,
+            );
+        }
+
+        fn free_set_open_callback(free_set_checkpoint: *CheckpointTrailer) void {
+            const grid = @fieldParentPtr(Grid, "free_set_checkpoint", free_set_checkpoint);
+            const callback = grid.callback.open;
+
+            {
+                assert(!grid.free_set.opened);
+                defer assert(grid.free_set.opened);
+
+                grid.free_set.open(.{
+                    .encoded = free_set_checkpoint.buffer[0..free_set_checkpoint.size],
+                    .block_addresses = free_set_checkpoint.block_addresses[0..free_set_checkpoint.block_count()],
+                });
+                assert((free_set_checkpoint.size > 0) == (grid.free_set.count_acquired() > 0));
```

Comment:
I wonder if we should add `free_set.pristine()`? I worry a little that `count_acquired` is O(N), while its enough to just find a single set bit (which probably going to be the first one even)

Reply:
I was a bit hesitant about empty, as I think we might want to check all of blocks, staging, and reservations. But I am probably overthinking this. And yes, the idea is to do an early return. Though I guess if we are checking staging, we’d still be O(N)?